python/pyiceberg/table/__init__.py

```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+        if field.field_id in self._updates:
+            raise ValueError(f"Cannot delete a column that has updates: {full_name}")
+
+        self._deletes.add(field.field_id)
+
         return self
 
-    def commit(self) -> None:
-        """Apply the pending changes and commit."""
-        new_schema = self._apply()
-        updates = [
-            AddSchemaUpdate(schema=new_schema, last_column_id=new_schema.highest_field_id),
-            SetCurrentSchemaUpdate(schema_id=-1),
-        ]
-        requirements = [AssertCurrentSchemaId(current_schema_id=self._schema.schema_id)]
+    def rename_column(self, path_from: Union[str, Tuple[str, ...]], new_name: str) -> UpdateSchema:
+        """Updates the name of a column.
+
+        Args:
+            path_from: The path to the column to be renamed.
+            new_name: The new path of the column.
+
+        Returns:
+            The UpdateSchema with the rename operation staged.
+        """
+        name_from = tuple(path_from.split(".")) if isinstance(path_from, str) else path_from
+        new_name = new_name.split(".")[-1]
 
-        if self._transaction is not None:
-            self._transaction._append_updates(*updates)  # pylint: disable=W0212
-            self._transaction._append_requirements(*requirements)  # pylint: disable=W0212
+        parent_path = name_from[:-1]
+        full_name_from = ".".join(name_from)
+
+        from_field = self._schema.find_field(full_name_from, self._case_sensitive)
+
+        if from_field.field_id in self._deletes:
+            raise ValueError(f"Cannot rename a column that will be deleted: {full_name_from}")
+
+        if updated := self._updates.get(from_field.field_id):
+            self._updates[from_field.field_id] = NestedField(
+                field_id=updated.field_id,
+                name=new_name,
+                field_type=updated.field_type,
+                doc=updated.doc,
+                required=updated.required,
+            )
         else:
-            table_update_response = self._table.catalog._commit_table(  # pylint: disable=W0212
-                CommitTableRequest(identifier=self._table.identifier[1:], updates=updates, requirements=requirements)
+            self._updates[from_field.field_id] = NestedField(
+                field_id=from_field.field_id,
+                name=new_name,
+                field_type=from_field.field_type,
+                doc=from_field.doc,
+                required=from_field.required,
             )
-            self._table.metadata = table_update_response.metadata
-            self._table.metadata_location = table_update_response.metadata_location
 
-    def _apply(self) -> Schema:
-        """Apply the pending changes to the original schema and returns the result.
+        if path_from in self._identifier_field_names:
+            self._identifier_field_names.remove(full_name_from)
+            self._identifier_field_names.add(f"{'.'.join(parent_path)}.{new_name}")
```

Comment:
Ah, I see why you wanted to parse `name_from`. The problem is that we can never parse full names using `split` because we don't know whether `.` is a delimiter or a character in a name. In this case, you should look up the parent's name by finding the parent's ID and looking it up in the ID to name index.

By the way, I think this fixes a problem in the Java implementation. The `newName` is a leaf name, not a full name.

Reply:
We had the same train of thought, I've updated this to:
```python
        if path_from in self._identifier_field_names:
            self._identifier_field_names.remove(path_from)
            self._identifier_field_names.add(f"{path_from[:-len(field_from.name)]}{new_name}")
```